### PR TITLE
Added support for remote racket

### PR DESCRIPTION
**Pull request by fakedrake**
When a racket session is created in a file visiting a remote server
via tramp the racket process is created in the remote server. If this
is the case the racket code running in the server is copied to the
remote server (if it is not there already). Said code is also modified
to accept connections from externa sources (security concern?).

This  is my solutiuon to issue #552 .

### CODE REVIEW

File: racket-repl.el [hunk 1]
```diff
@@ -526,79 +527,98 @@ WHAT-TO-RUN may be nil, meaning just a `racket/base` namespace."
...
+          ;; (e.g. `comint-input-ring' would lose input history).
+          (unless (eq major-mode 'racket-repl-mode)
+            (when noninteractive
+              (princ "{racket--repl-start}: (racket-repl-mode)\n"))
+            (racket-repl-mode)))
+      (file-error
+       (let ((kill-buffer-query-functions nil)
+             (kill-buffer-hook nil))
+         (kill-buffer)) ;don't leave partially initialized REPL buffer
+       (message "Could not connect to REPL server: %s"
```
Comment (greghendershott) [inline racket-repl.el:619]: I think this is the format string the bytecode compiler is complaining about? https://github.com/greghendershott/racket-mode/actions/runs/1117388372

File: racket-cmd.el [hunk 2]
```diff
@@ -77,39 +77,74 @@ that Racket Mode can find its own run.rkt file.")
...
-                            (funcall racket-adjust-run-rkt racket--run.rkt)
-                            "--auth"
-                            (setq racket--cmd-auth (format "token-%x" (random)))
-                            (if (and (boundp 'image-types)
-                                     (fboundp 'image-type-available-p)
-                                     (or (and (memq 'svg image-types)
-                                              (image-type-available-p 'svg))
-                                         (and (memq 'imagemagick image-types)
-                                              (image-type-available-p 'imagemagick))))
+  (let ((cmd (list racket-program
```
Comment (greghendershott) [inline racket-cmd.el:114]: This assumes that `racket-program` will work on both local and remote -- e.g. that `racket` is good on both, as opposed to `/path/to/specific/version/of/racket` on one but not the other. That assumption will probably be fine for most people, so maybe it's fine to punt on this, as opposed to adding Yet Another Customization Variable? What do you think?

File: racket-custom.el [hunk 3]
```diff
@@ -138,6 +138,15 @@ and `racket-repl-documentation' should look for the search page.
   :safe (lambda (val) (or (stringp val) (eq val 'local)))
   :group 'racket)
 
+(defcustom racket--rkt-remote-source-dir
```
Comment (greghendershott) [inline racket-custom.el:141]: Until I read your out-of-band comment, it wasn't clear to me (I was dense I guess) that this feature would involve copying the Racket Mode back end Racket code to the remote server. So I think I would probably rewrite the prose for this doc string, to make that clearer for end users. But that's something I'm happy to do myself in a separate commit.

File: racket/repl.rkt [hunk 4]
```diff
@@ -128,8 +128,8 @@
 (define listener
   (tcp-listen 0  ;choose port dynamically, a.k.a. "ephemeral" port
               64
-              #f ;reuse? not recommended for ephemeral ports
-              "127.0.0.1"))
+              #t ;reuse? not recommended for ephemeral ports
```
Comment (greghendershott) [inline racket/repl.rkt:131]: Regarding the security concern: I think it might be fine if the remote back end is started using a newly added command-line argument such as `--listen-to` (defaulting to 127.0.0.1) and that value is used here. Unfortunately I guess one flaw to supplying the "what is my IP?" value automatically is the usual complication of NATs. So not yet sure best thing to do here.

**Comment by greghendershott** [comment]
I made some inline comments. I hope they're helpful!

One more comment that I didn't know where to put inline: Probably this will only work reliably, especially when working with a mix of local and remote source files, when each source file has its own unique REPL session.

That is, when the value of `racket-repl-buffer-name-function` is [`racket-repl-buffer-name-unique`](https://www.racket-mode.com/#racket_002drepl_002dbuffer_002dname_002dunique). (Which is not the default value. Instead by default a single REPL is shared, which is the pre-multi-REPLs behavior.)

I think that's fine; we'd just need to document that value needs to be used. (I'd be happy to write that prose.)

File: racket-cmd.el [hunk 5]
```diff
@@ -77,39 +77,74 @@ that Racket Mode can find its own run.rkt file.")
...
-     :filter          #'racket--cmd-process-filter)))
+                     "--do-not-use-svg"))))
+    (cl-letf* (((default-value 'process-environment) process-environment)
+               ((default-value 'exec-path)           exec-path))
+      (message "Running racket: %s" cmd)
+      (let* ((buf (get-buffer-create (concat " *" racket--cmd-process-name "*")))
+             (errbuf (get-buffer-create (concat " *" racket--cmd-process-name ":stderr*")))
+             (is-remote (and default-directory (file-remote-p default-directory)))
+             (make-process-fn
+              (if is-remote #'tramp-sh-handle-make-process #'make-process))
```
[the remark this answers is not in the document]
Comment (fakedrake) [reply racket-cmd.el:136]: You are probably right, good suggestion. I didn't test it with older emacsen.

File: racket/repl.rkt (hunk 4 above)
Comment (fakedrake) [reply racket/repl.rkt:131]: That would be nice, we need a way to find out the IP from elisp

File: racket-custom.el (hunk 3 above)
Comment (fakedrake) [reply racket-custom.el:141]: You are right

**Comment by greghendershott** [comment]
Please enjoy your vacation and ignore the rest of this comment until you're back!

---

Is your vacation over? Really? :smile:

You:

> Per-project repl would be nice.

That does already exist.

Me, earlier:

> One more comment that I didn't know where to put inline: Probably this will only work reliably, especially when working with a mix of local and remote source files, when each source file has its own unique REPL session.
>
> That is, when the value of racket-repl-buffer-name-function is racket-repl-buffer-name-unique. (Which is not the default value. Instead by default a single REPL is shared, which is the pre-multi-REPLs behavior.)
>
> I think that's fine; we'd just need to document that value needs to be used. (I'd be happy to write that prose.)

Actually it's not quite this simple.

A mix of local and remote files could still cause problems, even with a dedicated REPL for each.

Racket Mode expects there to be only a _single_ back end. That single back end supports zero or more REPLs.

The back end is used for quite a few commands that don't involve a REPL.

It's not clear to me what it would mean for Racket Mode to be able to juggle multiple back end processes, at all, much less a mix of local and remote.

Initially, that sounds like a buggy hairball. But, while you continue to enjoy your vacation, maybe I'll mull that over.

(Worst case, we could just advise people: "Do `M-x racket-stop-back-end` when switching between local and remote files". That's jankier than I usually like, but maybe it would be acceptable-ish (??) for your workflow and that of other people using this feature.)

**Comment by fakedrake** [comment]
So one more problem with this PR is that it does not play well with `racket-open-require-path`.

**Comment by greghendershott** [comment]
Over the past week, I've been working on a branch to support multiple back ends (any mix of local and remote). As well as incorporating the gist of your PR. I've been encountering a number of issues, and trying to tackle them one by one.

This has been somewhat slow going as I've been marinading myself in how tramp works etc.

> So one more problem with this PR is that it does not play well with `racket-open-require-path`.

- Yep. That directly invokes `racket-program` locally. I think it could be adapted instead to do that remotely.

- A similar issue is some of the documentation commands do a `raco doc` locally. Invoking that remotely isn't exactly the answer as that will try to open a browser remotely.

- Also there are a couple things like `racket-xp-mode` where `check-syntax` returns pathnames -- for documentation or jump-to-definition. When run remotely, those will be paths on the remote. The commands would need to change those into tramp filenames usable locally.

**Comment by greghendershott** [comment]
I've continued to work on this in the [`multi-back-end`](https://github.com/greghendershott/racket-mode/tree/multi-back-end). Although I suspect there are more problems I don't yet know about, I've addressed the ones listed above -- except for invoking `raco doc` remotely.

I don't know if you're willing or able to try that branch, but if so, and if you have any comments, please let me know.

**Comment by greghendershott** [comment]
Note: Working on this over several weeks, I accumulated some changes for multi/remote back ends, as well as to enhance `racket-describe`.

Why the mix: As I was dog-fooding this, I realized that `racket-describe` was important for using remote back ends, and, leaning on that more meant it should be improved. It was easiest to make the improvements during the experience of working with a remote back end, i.e. while dog-fooding.

However the `multi-back-ends` branch got into a state where it had dozens of commits, mixing both themes. This was messy. Also I wanted to the option to merge the `racket-describe` changes, sooner.

So recently I split this out:

https://github.com/greghendershott/racket-mode/tree/better-describe is squashed commits only about improving `racket-describe`. I expect to merge this first.

https://github.com/greghendershott/racket-mode/tree/rebased-multi-back-end is squashed and has only the changes related to multi back ends. I expect to merge this afterwards; periodically it needs to rebased onto `better-describe` (with the idea that `better-describe` will be merged to `master` first).

**Comment by greghendershott** [comment]
To update this for anyone else following along:

- The `better-describe` branch was merged some months ago.
- Although I had set it aside for some weeks (mainly to work on the `hash-lang` branch to keep up with some developments from Racket core team in that area) I've continued to work on the `rebased-multi-back-end` branch, and @fakedrake has continued to use it and offer feedback. I think it is "feature complete".
- One important point (to me) is that if someone isn't using multiple back ends, they don't need to do any configuration. That simple scenario is supposed to "just work" like before; i.e. "do no harm" for those who don't need or care about this.
- I think `rebased-multi-back-end` might be ready to merge, depending on any more feedback.
